lib/Segments/DynamicSegments/Filters/WooCommerceNumberOfOrders.php

```diff
@@ -0,0 +1,59 @@
+<?php
+
+namespace MailPoet\Segments\DynamicSegments\Filters;
+
+use MailPoet\Entities\DynamicSegmentFilterEntity;
+use MailPoet\Entities\SubscriberEntity;
+use MailPoetVendor\Carbon\Carbon;
+use MailPoetVendor\Doctrine\DBAL\Query\QueryBuilder;
+use MailPoetVendor\Doctrine\ORM\EntityManager;
+
+class WooCommerceNumberOfOrders implements Filter {
+  const ACTION_NUMBER_OF_ORDERS = 'numberOfOrders';
+
+  /** @var EntityManager */
+  private $entityManager;
+
+  public function __construct(EntityManager $entityManager) {
+    $this->entityManager = $entityManager;
+  }
+
+  public function apply(QueryBuilder $queryBuilder, DynamicSegmentFilterEntity $filter): QueryBuilder {
+    global $wpdb;
+    $subscribersTable = $this->entityManager->getClassMetadata(SubscriberEntity::class)->getTableName();
+    $filterData = $filter->getFilterData();
+    $type = $filterData->getParam('number_of_orders_type');
+    $count = $filterData->getParam('number_of_orders_count');
+    $days = $filterData->getParam('number_of_orders_days');
+
+    $date = Carbon::now()->subDays($days);
+
+    $queryBuilder->innerJoin(
+      $subscribersTable,
+      $wpdb->postmeta,
+      'postmeta',
+      "postmeta.meta_key = '_customer_user' AND $subscribersTable.wp_user_id=postmeta.meta_value"
+    )->leftJoin(
+      'postmeta',
+      $wpdb->posts,
+      'posts',
+      'posts.ID = postmeta.post_id AND posts.post_date >= :date AND postmeta.post_id NOT IN ( SELECT id FROM ' . $wpdb->posts . ' as p WHERE p.post_status IN ("wc-cancelled", "wc-failed"))'
```

Comment:
I am not sure about this subselect. I am not very good at MySQL, but I was told to avoid subselects as much as possible because they tend to be slow. The reason, I was told, is that MySQL needs to evaluate the subselect for each row. That can be slow But I am just asking here, did you test this? or do you know much about the final query's performance?

Reply:
I'm not a MySQL expert as well, but I think that this query in general is not great. That being said, I'm not sure if there is much we can do about it at the moment in this particular PR. We are already using exactly the same subquery in the other two WooCommerce segments:

https://github.com/mailpoet/mailpoet/blob/5b45bdac1e0ba38e99d20aad7f6883c8863ba892/lib/Segments/DynamicSegments/Filters/WooCommerceCategory.php#L63
https://github.com/mailpoet/mailpoet/blob/5b45bdac1e0ba38e99d20aad7f6883c8863ba892/lib/Segments/DynamicSegments/Filters/WooCommerceProduct.php#L39

That is why I'm using it here, and I did not test it. I'm sure this query will be slow for large databases, but I don't think the subquery is the worst part. Probably the worst part is the inner join with wp_postmeta, which uses meta_value since this field is not indexed:

```
mysql> explain SELECT DISTINCT wp_mailpoet_subscribers.id as inner_subscriber_id FROM wp_mailpoet_subscribers INNER JOIN wp_postmeta postmeta ON postmeta.meta_key = '_customer_user' AND wp_mailpoet_subscribers.wp_user_id=postmeta.meta_value LEFT JOIN wp_posts posts ON posts.ID = postmeta.post_id AND posts.post_date >= '2021-04-04 18:24:20' AND postmeta.post_id NOT IN ( SELECT id FROM wp_posts as p WHERE p.post_status I
N ("wc-cancelled", "wc-failed")) GROUP BY inner_subscriber_id HAVING COUNT(posts.ID) > 1;
+----+--------------------+-------------------------+------------+-----------------+-------------------------------------------------------------------------------------------------------+------------+---------+-------------------------------+------+----------+----------------------------------------------+
| id | select_type        | table                   | partitions | type            | possible_keys                                                                                         | key        | key_len | ref                           | rows | filtered | Extra                                        |
+----+--------------------+-------------------------+------------+-----------------+-------------------------------------------------------------------------------------------------------+------------+---------+-------------------------------+------+----------+----------------------------------------------+
|  1 | PRIMARY            | postmeta                | NULL       | ref             | meta_key                                                                                              | meta_key   | 767     | const                         |   11 |   100.00 | Using where; Using temporary; Using filesort |
|  1 | PRIMARY            | wp_mailpoet_subscribers | NULL       | ref             | PRIMARY,email,unsubscribe_token,wp_user_id,updated_at,status_deleted_at,last_subscribed_at,link_token | wp_user_id | 9       | wordpress.postmeta.meta_value |    1 |   100.00 | Using where; Using index                     |
|  1 | PRIMARY            | posts                   | NULL       | eq_ref          | PRIMARY                                                                                               | PRIMARY    | 8       | wordpress.postmeta.post_id    |    1 |   100.00 | Using where                                  |
|  2 | DEPENDENT SUBQUERY | p                       | NULL       | unique_subquery | PRIMARY                                                                                               | PRIMARY    | 8       | func                          |    1 |    20.00 | Using where                                  |
+----+--------------------+-------------------------+------------+-----------------+-------------------------------------------------------------------------------------------------------+------------+---------+-------------------------------+------+----------+----------------------------------------------+
4 rows in set, 1 warning (0.01 sec)
```

If we are concerned about the performance of our database queries, and I think we should be, I suggest we create a task to investigate what are our options for all WooCommerce segments as I don't think there is an easy fix here. What do you think?

Reply:
> If we are concerned about the performance of our database queries, and I think we should be, I suggest we create a task to investigate what are our options for all WooCommerce segments as I don't think there is an easy fix here. What do you think?

You're right. I didn't realise this is like that in other segments as well. So leave it for now, but yeah, we should revisit that later. I think it will be part of [these tasks](https://wordpress.com/read/feeds/113756048/posts/3267798944), right?

Reply:
I think it is related (here is the link to the task https://mailpoet.atlassian.net/browse/MAILPOET-3557), but I'm not sure it is the same thing. The task that I created is focused on describing how we are currently interacting with WooCommerce data and doesn't focus on the performance of the queries. Maybe we could create a related task to investigate our options to improve the performance of the queries used in the WooCommerce segments?

Reply:
Sure we can do that. 

I was thinking that eventually, we want to switch from the SQL queries to Woocommerce functions calls. That would solve that. But I am not sure if that will happen or when, so creating a ticket is an option.

Reply:
Ticket created here: https://mailpoet.atlassian.net/browse/MAILPOET-3571